### ISSUE TITLE
Paramterized mappings not working using VB.NET and Profiles

### ISSUE THREAD

**Issue by CalinLeafshade** [AutoMapper/AutoMapper#1748]
I've defined the following Profile:

```VB
Public Class LearningPlanViewModelProfile
    Inherits Profile

    Protected Overrides Sub Configure()


        ' This will be parameterised in projections
        Dim userID As Integer = 0

        CreateMap(Of LearningPlan, LearningPlanViewModel)().
            ForMember(Function(dest) dest.TAPs,
                      Sub(opts) opts.MapFrom(Function(src) src.LearningPlanRules.
                                                 Select(Function(y) y.ContentLabel).
                                                    SelectMany(Function(y) y.ContentLabelTAPs).
                                                    Where(Function(y) y.TAP IsNot Nothing).
                                                    Select(Function(y) y.TAP).Distinct()))

        CreateMap(Of TAP, LearningPlanTapViewModel)().
            ForMember(Function(src) src.Completion,
                      Sub(opts) opts.MapFrom(Function(src) src.UserTAPs.
                                                 Where(Function(ut) ut.UserID = userID).
                                                 Select(Function(ut) ut.PercentComplete).
                                                 DefaultIfEmpty(0).
                                                 FirstOrDefault()))

    End Sub

End Class
```

And I'm mapping like this:

```VB
Dim plans = Await _lpQueryService.
                    GetAllLearningPlansForUserQuery(Principal.UserID, Principal.CompanyScopeId).
                    ProjectTo(Of LearningPlanViewModel)(_mapper.ConfigurationProvider, New Dictionary(Of String, Object) From {{"userID", Principal.UserID}}).ToArrayAsync()
```

The map always returns the results as if the userID param were 0. If I change the default value of that parameter then it uses that value. It doesn't seem to be correctly injecting the new value.

**Comment by lbargaoanu** [comment]
Did you try to set the parameter to a constant and then look for that value in the generated query? Because I did that and it looks good. The final result doesn't, but that's another story.

**Comment by CalinLeafshade** [comment]
If I change the paramter to a constant (i.e replace Principal.UserID with a numer) it still shows as 0 in the query so it seems like it's not injecting the parameter at all.

**Comment by jbogard** [comment]
The next thing I would do is look at the debug viewer of the expression that AutoMapper builds. You can do this if you capture in a variable the expression right BEFORE you call ToList or ToArray:

```
Dim plansQuery = _lpQueryService.
                    GetAllLearningPlansForUserQuery(Principal.UserID, Principal.CompanyScopeId).
                    ProjectTo(Of LearningPlanViewModel)(_mapper.ConfigurationProvider, New Dictionary(Of String, Object) From {{"userID", Principal.UserID}})
Dim plans = Await plansQuery.ToArrayAsync()
```
If you break at the first line, and look in the debug window, the IQueryable will contain an Expression property. There's a debug viewer that shows *exactly* what was built. That should show you what EF is receiving, and translating into a query.

**Comment by CalinLeafshade** [comment]
The relevant part of the expression seems to be this:

```
.Lambda #Lambda15<System.Func`2[MyProject.UserTAP,System.Boolean]>(MyProject.UserTAP $ut) {
    (System.Boolean)($ut.UserID == .Constant<MyProject.LearningPlanViewModelProfile+_Closure$__1-0>(MyProject.LearningPlanViewModelProfile+_Closure$__1-0).$VB$Local_userID)
}
```

That's using $VB$Local_userID as the value that its checking against. I'm not sure if this is correct or if it should be something else

**Comment by lbargaoanu** [comment]
Instead of "userID" try "$VB$Local_userID" :) Or make a repro in C#.

**Comment by lbargaoanu** [comment]
Of course we can fix this, but first we should find the cause.

**Comment by CalinLeafshade** [comment]
I tried a few variants of that string and none seem to work. Looks like it might be a peculiarity of the VB compiler.

**Comment by jbogard** [comment]
Well that's a closure, so it's capturing the value in a class. C# does the exact same thing, but its closure looks a bit different. @lbargaoanu I wonder if there's metadata attached to that generated class that we could look for?

**Comment by lbargaoanu** [comment]
@jbogard Maybe. But I would guess it's a simple naming convention.

**Comment by lbargaoanu** [comment]
Yes, it's not the name of the field, it's [the name of the class](https://github.com/AutoMapper/AutoMapper/blob/57ff9e9f297501f9532bf9e6cd27d932108c40c6/src/AutoMapper/QueryableExtensions/ExpressionBuilder.cs#L267).

**Comment by lbargaoanu** [comment]
For the class. For fields is just the naming convention.

**Comment by CalinLeafshade** [comment]
@lbargaoanu Could I have a binary to test the PR?

**Comment by lbargaoanu** [comment]
Try to build the [branch](https://github.com/lbargaoanu/AutoMapper/tree/VBProjectToParameters). You can build only AutoMapper.Net4, for the full .NET.

**Comment by CalinLeafshade** [comment]
Yep, that seems to work as expected. Good work guys. This can be closed from my perspective once the PR is merged in.
